### ISSUE TITLE
[spdk-abi] Tracking changes in debug mode

### ISSUE THREAD

**Issue by tomzawadzki** [spdk/spdk#3096]
# Sighting report

While developing changes[[1](https://review.spdk.io/gerrit/c/spdk/spdk/+/19281/)] to `check_so_deps` test script allowing developers to use it more easily, it was found that API is tracked only with `--enable-debug`.
This is not the right choice, as what actually counts is the release mode (`--disable-debug`).

Fortunately the difference occurs only in single instance right now[[2](https://github.com/spdk/spdk/blob/e72f6652c492a9e62bab06892d854462fb56c1fa/include/spdk/sock.h#L52)].

I have not analyzed past SPDK versions to see if similar changes in debug code resulted in version bumps.

## Expected Behavior

Only changes done with `--disable-debug` should be versioned for SPDK libs.

## Current Behavior

Changes in debug mode might affect testing with abidiff and require version bumps.

## Possible Solution

Two options are proposed, in both the spdk/spdk-abi repo would be updated with API built in release mode.

Option 1)
Is to remove any instance of `#ifdef DEBUG` in `./include/spdk/*` and add verification that it does not appear in the future (`check_format.sh`). An exception would have to be added when equivalent #else is present see `include/spdk/log.h`.
For the case in `sock.h` the `curr_list` would always be present and only operations on it would be performed in debug mode as is right now. 

Option 2)
Allow use of `#ifdef DEBUG` without equivalent `#else`. The test will need to make sure that SPDK is built in release mode before comparison and by extension will require developers to rebuild.

**Comment by benlwalker** [comment]
I believe the ABI of the debug build must exactly match the ABI of the release build (i.e. option 1) for a few reasons:

* For shared library builds, it may be useful to drop in a debug library temporarily for some deployments. I know we're mostly discussing adding symbols here so the debug library is most likely compatible, but I think we should adopt a strong stance that they must be identical just to avoid any future issues.
* It's really convenient for developers to be able to use debug builds at their desk and do ABI checking as they build
* It's fairly easy to workaround any use of `#ifdef DEBUG` by just adding an `#else` like you outlined

So I vote option 1.

**Comment by jimharris** [comment]
@karlatec - could your team run some nvme-tcp perf tests with https://review.spdk.io/gerrit/c/spdk/spdk/+/19752?  Both initiator and target side - high throughput tests.

Thanks,

-Jim

**Comment by jimharris** [comment]
> Sure, no problem. You can find the results [here](https://performance.spdk.io/d/4tDkVyd4k/spdk-nvme-of-tcp-performance?orgId=1&refresh=5s&var-cpus=4&var-qd=128&var-rw=randread&var-manual_builds=true) (just make sure "Build annotations" are enabled, seems this selection is not copied in URL). (they're going to be highlighted with yellow dotted lines).
> 
> FYI: There's some issue on this particular performance setup right now and performance dropped around end of August. I'm working to get it fixed. When I do I'm going to re-run tests for this commit.

It looks to me like there isn't any measurable difference with this patch.  Do you agree?

**Comment by karlatec** [comment]
Agreed. I've re-run the build after fixing the test system. There are new results available in the dashboard.

**Comment by tomzawadzki** [comment]
[Community call] Since the https://github.com/spdk/spdk/commit/49be050e1cae48f3aa1ae736141189e84bf647d2 was merged, the bug is resolved. Addressing the testing of future changes is all that remains, so this issue will be relabeled enhancement.

**Comment by jimharris** [comment]
My patch to remove DEBUG from spdk_sock_request was merged, so I'm going to unassign this issue from me for now.

**Comment by tomzawadzki** [comment]
[Bug scrub] The current spdk-abi test should be switched to compare release builds of SPDK, and new test added that would compare patch under test with itself between two types of builds.

**Comment by mikeBashStuff** [comment]
Just to leave a note that this is WIP, but will try to push some patches this week.
